### ISSUE TITLE
Add option to enable libnpp?

### ISSUE THREAD

**Issue by msmollin** [rdp/ffmpeg-windows-build-helpers#199]
One of the things I'd like to be able to do is hardware-accelerate video resizing during transcoding, which is available using the (nonfree) library libnpp if you have a compatible Nvidia graphics card. I would submit a patch for this, but I could use a little guidance / information on how to do that, because libnpp requires signing up with Nvidia's developer program and then downloading the NVENC SDK (which includes the primitives SDK) in order to get the libraries. 

Users then have to add extra cflags and ldflags options to the main FFMPEG configure options list to point to the location you downloaded the SDK, unless you put the SDK somewhere in the path. So it sounds like I would need to add the following:

1) A `--ldflags` parameter similar to the existing cflags parameter,
2) An option to enable libnpp separately since its likely most users won't want this
3) An if check to check to see if that option is available inside the `build_ffmpeg()` function

Anything I missed? I could imagine we could simplify a bit and hardcode a path for the the first option somewhere relative to the script so as to keep it simpler, but I don't know if that would frustrate some users? 

Anything I missed? I imagine we could simplify a bit and hardcode a path for the the first option somewhere relative to the script so as to keep it simpler, but I don't know if that would frustrate some users?

**Comment by hydra3333** [comment]
https://ffmpeg.zeranoe.com/forum/viewtopic.php?f=3&t=858&start=40#p11902

**Comment by Collisionc** [comment]
I made a fork with libnpp support a few months ago, I'm not the best at scripting but it produces a ffmpeg binary (when https://github.com/rdp/ffmpeg-windows-build-helpers/issues/200 is resolved, that is) with cuvid/libnpp support. It requires nonfree to be enabled since libnpp is nonfree.

https://github.com/Collisionc/ffmpeg-windows-build-helpers

It will require CUDA Toolkit 8.0 - https://developer.nvidia.com/cuda-downloads installed on the pc using ffmpeg. Unfortunately there doesn't seem to be a way to statically compile scale_npp support on Windows, and this is the only way to get the shared libraries so ffmpeg doesn't complain about missing .dlls when loaded.

One note, don't waste your time with cuvid... it's buggy and unreliable on anything other than h264/hevc.  DXVA2 is nearly as fast and doesn't crash.

**Comment by mtrandl** [comment]
> It will require CUDA Toolkit 8.0
Does this need to be compiled or just the headers from the source tarball?

**Comment by Collisionc** [comment]
Compiling it does not require the toolkit, running it will require the .dlls in windows path

**Comment by hydra3333** [comment]
> One note, don't waste your time with cuvid... it's buggy and unreliable on anything other than h264/hevc. DXVA2 is nearly as fast and doesn't crash.
That's what I found ... however :-

It seems possible that could actually be related to be an nvidia driver issue ?
https://ffmpeg.zeranoe.com/forum/viewtopic.php?f=3&p=11954#p11954
http://forum.videohelp.com/threads/382164-To-GPU-process-or-not-to-GPU-process-450fps-HD-SD?p=2475424&viewfull=1#post2475424

**Comment by Collisionc** [comment]
Yeah, that sounds like exactly what I ran into.  I also ran into some issues with h264 where cuvid would just drop about 5 seconds worth of frames randomly in certain files (always the same frames), while dxva2 was fine.

**Comment by Collisionc** [comment]
Semi-related, have you tried using cuvid/nvdec to downscale resolution instead of scale_npp ?  It was added to the ffmpeg github sometime this week, and it is fast.

Downscaling 3840x1920 h264 ---> 1920x960 h264 - Using 1080 GTX with the following lines:
Using ffmpeg built in filter to scale:
ffmpeg.exe -c:v h264_cuvid -i testfile.mp4 -vf scale=1920:1080 -c:v nvenc_h264 -y testoutput.mp4.
Average FPS: 70

scale_npp:
ffmpeg.exe -c:v h264_cuvid -i testfile.mp4 -vf hwupload_cuda,scale_npp=1920:1080:format=same:interp_algo=super,hwdownload,format=same -c:v nvenc_h264 -y testoutput.mp4
Average FPS: 125

resize with cuvid/nvdec:
ffmpeg.exe -c:v h264_cuvid -resize 1920x1080 -i testfile.mp4 -c:v nvenc_h264 -y testoutput.mp4
Average FPS: 195

The quality of the output is near-identical and it also supports resizing + cropping -- I don't believe scale_npp supports cropping.  

h264_cuvid AVOptions:
  -deint             <int>        .D.V.... Set deinterlacing mode (from 0 to 2) (default weave)
     weave                        .D.V.... Weave deinterlacing (do nothing)
     bob                          .D.V.... Bob deinterlacing
     adaptive                     .D.V.... Adaptive deinterlacing
  -gpu               <string>     .D.V.... GPU to be used for decoding
  -surfaces          <int>        .D.V.... Maximum surfaces to be used for decoding (from 0 to INT_MAX) (default 25)
  -drop_second_field <boolean>    .D.V.... Drop second field when deinterlacing (default false)
  -crop              <string>     .D.V.... Crop (top)x(bottom)x(left)x(right)
  -resize            <string>     .D.V.... Resize (width)x(height)


btw: With the newest nvidia drivers - 378.92 - I haven't run into problems with cuvid randomly crashing or dropping frames -- yet.

**Comment by hydra3333** [comment]
Thanks !
Could you attach a file or something showing the changes to the RDP script that you made to build with CUDA (cuvid/nvdec) and with npp ? 
(I'd like to add my opencl changes and make that my new baseline).

**Comment by Collisionc** [comment]
https://github.com/rdp/ffmpeg-windows-build-helpers/compare/master...Collisionc:master

Is the difference between mine and RDPs to enable npp. Unfortunately, there is no easy automated way to grab the source files for npp from nvidia since you have to download the entire 1.2 GB installer, so I just added the files to the repo. It's always nonfree as well, even with LGPL.

If you want to add the libnpp source files yourself, you can download them here - https://developer.nvidia.com/cuda-downloads

**Comment by hydra3333** [comment]
Thanks !

**Comment by hydra3333** [comment]
given you built with cuda recently, can you confirm this ?
https://ffmpeg.org/pipermail/libav-user/2017-January/010042.html which says
> On Thu, Jan 12, 2017 at 8:19 PM, Morten Fugl <morten at sportcaster.dk> wrote:
> > When changing ffmpeg to support nvidia hw decoding, I followed this guide:
> > https://developer.nvidia.com/ffmpeg
> > The guide says "Copy the NVIDIA video header files to the respective include
> > folders", but which files should I copy to which folders?
> 
> If you use a recent Git master version of ffmpeg, you don't need to
> copy any headers anymore, cuda/cuvid decoding and nvenc encoding is
> supported out of the box.
> - Hendrik

i,e. no longer any need to download nvidia sdk/headers and copy them to spefiic folders so that ffmpeg builds cuvid (cida) ?

**Comment by Collisionc** [comment]
correct, cuvid/nvdec is included in ffmpeg source now.  libnpp is not part of that, though.

**Comment by hydra3333** [comment]
OK thank you.  

Now I can simply add this to the script's build commandline in rdp's script (along with my script mods to build also openCL for nvidia) so that nvenc and cuda/cuvid all get built automagically !
>   --enable-nvenc --enable-cuda --enable-cuvid

@rdp would it be possible for you to add that as a standard default in your script ?

Then I'm interested enough to start testing, a la this
https://github.com/rdp/ffmpeg-windows-build-helpers/issues/199#issuecomment-288144455
and see whether I get the zillions of decoding errors I got last time.

**Comment by rdp** [comment]
Appears the default is to have them on so I guess we get it "for free"?
```

$ ./configure --help | grep nvenc
  --disable-nvenc          disable Nvidia video encoding code [autodetect]
$ ./configure --help | grep cuda
  --disable-cuda           disable dynamically linked Nvidia CUDA code [autodetect]
$ ./configure --help | grep cuvid
  --disable-cuvid          disable Nvidia CUVID support [autodetect]
```

**Comment by Collisionc** [comment]
yeah they should be auto detected, I think you can only disable them.

**Comment by hydra3333** [comment]
My goodness. Nice !

**Comment by schmidthubert** [comment]
>ffmpeg.exe -c:v h264_cuvid -resize 1920x960 -i testfile
Unfortunatly there's no "-resize 1920x-1" as in "-vf scale=1920:-1" to keep aspect ratio so no good yet for batch jobs etc ... =:-(

**Comment by schmidthubert** [comment]
Collisionc fork seems DEAD. But luckily there's https://github.com/jb-alvarado/media-autobuild_suite and #183.. So actually this could be closed as well. No npp "issues" anymore, thank god ... :)

**Comment by rdp** [comment]
See also https://stackoverflow.com/questions/41870137/ffmpeg-error-libnpp-not-found-in-windows and https://github.com/rdp/ffmpeg-windows-build-helpers/issues/249 https://github.com/m-ab-s/media-autobuild_suite/blob/dfe7cb2fa15df64ce8d918654ac6b008178613e8/build/media-suite_helper.sh#L1952 needs some local dependencies?  Can you even run those in linux?  wine? LOL
